### PR TITLE
[IMP] pos_*: introducing refund with razorpay payment terminal

### PR DESCRIPTION
**Pull request by paan-odoo**
pos_*: point_of_sale, pos_razorpay

Before this commit:
=========
- There is no refund option with the Razorpay payment terminal. Users need to refund orders paid with the Razorpay payment method using cash or another.
- We were taking request timeout from the config parameter the Razorpay API request.

After this commit:
==========
- Users will have the option to refund Razorpay terminal-based paid orders using
  Razorpay.
- A refund button will be provided on the terminal refund payment line. Clicking
  on this button will initiate the terminal refund process, and the order will
  be validated after the successful refund process.
- We will use static request timeout for the Razorpay API request.

task-4061153

### CODE REVIEW

File: addons/pos_razorpay/static/src/app/payment_razorpay.js [hunk 1]
```diff
@@ -97,13 +135,37 @@ export class PaymentRazorpay extends PaymentInterface {
...
-        const data = {
-            amount: line.amount,
-            referenceId: localStorage.getItem("referenceId"),
-        };
-        return this._call_razorpay(data, "razorpay_make_payment_request").then((data) => {
-            return this._razorpay_handle_response(data);
-        });
+        if (order._isRefundOrder()) {
+            const data = {
+                amount: -line.amount,
```
Comment (paso-odoo) [inline addons/pos_razorpay/static/src/app/payment_razorpay.js:140]:
```suggestion
                amount: Math.abs(line.amount),
```

File: addons/pos_razorpay/static/src/app/payment_razorpay.js [hunk 2]
```diff
@@ -97,13 +135,37 @@ export class PaymentRazorpay extends PaymentInterface {
...
+            const response = await this._check_payment_status(line);
+            if (response?.settlementStatus === "SETTLED") {
+                return this._call_razorpay(data, "razorpay_make_refund_request").then((data) => {
+                    return this._razorpay_handle_refund_response(data);
+                });
+            } else {
+                return this._call_razorpay(data, "razorpay_make_void_request").then((data) => {
+                    return this._razorpay_handle_refund_response(data);
+                });
+            }
```
Comment (paso-odoo) [inline addons/pos_razorpay/static/src/app/payment_razorpay.js:153]: use the same method and pass refund_type as refund or void
```suggestion
            if (response?.settlementStatus === "SETTLED") {
                data.refund_type = 'refund'
            }
            else {
                data.refund_type = 'void'
            }   
            return this._call_razorpay(data, "razorpay_make_refund_request").then((data) => {
                    return this._razorpay_handle_refund_response(data);
            });
```

File: addons/pos_razorpay/models/pos_payment_method.py [hunk 3]
```diff
@@ -15,6 +15,64 @@ class PosPaymentMethod(models.Model):
...
+                'nameOnCard': response.get('nameOnCard'),
+                'acquirerCode': response.get('acquirerCode'),
+                'postingDate': response.get('postingDate'),
+            }
+        default_error_msg = _('Razorpay POS payment request expected errorCode not found in the response')
+        error = response.get('errorMessage') or default_error_msg
+        return {'error': str(error)}
+
+    def razorpay_make_refund_request(self, data):
+        razorpay = RazorpayPosRequest(self)
```
Comment (paso-odoo) [inline addons/pos_razorpay/models/pos_payment_method.py:47]:
```
if data.get('refund_type') == 'refund':
   //  make separate logic for refund
else:
   //  make separate logic for void

// Keep common things for both requests
```

Do something like this

File: addons/pos_razorpay/models/pos_payment_method.py [hunk 4]
```diff
@@ -15,6 +15,64 @@ class PosPaymentMethod(models.Model):
...
+                'paymentMode': response.get('paymentMode'),
+                'paymentCardType': response.get('paymentCardType'),
+                'paymentCardBrand': response.get('paymentCardBrand'),
+                'nameOnCard': response.get('nameOnCard'),
+                'acquirerCode': response.get('acquirerCode'),
+                'postingDate': response.get('postingDate'),
+            }
+        default_error_msg = _('Razorpay POS payment request expected errorCode not found in the response')
+        error = response.get('errorMessage') or default_error_msg
+        return {'error': str(error)}
```
Comment (paso-odoo) [inline addons/pos_razorpay/models/pos_payment_method.py:44]:
```suggestion
```

remove this method and add logic in a single method to reduce redundant code

File: addons/pos_razorpay/models/razorpay_pos_request.py [hunk 5]
```diff
@@ -29,7 +34,8 @@ def _call_razorpay(self, endpoint, payload):
         :return The JSON-formatted content of the response.
         :rtype: dict
         """
-        endpoint = f'{self._razorpay_get_endpoint()}{endpoint}'
```
Comment (paso-odoo) [inline addons/pos_razorpay/models/razorpay_pos_request.py:32]: use this method to get endpoints

**Review by paso-odoo — CHANGES_REQUESTED** [review]

File: addons/point_of_sale/static/src/app/models/pos_order.js [hunk 6]
```diff
@@ -861,6 +861,20 @@ export class PosOrder extends Base {
...
+            return false;
+        } else {
+            return this.payment_ids.some((pl) => {
+                if (pl.payment_method_id.use_payment_terminal) {
+                    return pl.payment_status !== "done";
+                } else {
+                    return false;
+                }
+            });
+        }
```
Comment (rare-odoo) [inline addons/point_of_sale/static/src/app/models/pos_order.js:875]: You can avoid the `else` section
```
if (this._isRefundOrder()) {
    return this.payment_ids.some((pl) => {
        if (pl.payment_method_id.use_payment_terminal) {
            return pl.payment_status !== "done";
        }
        return false;
    });
}
return false;
```

File: addons/pos_razorpay/static/src/app/screens/payment_screen/payment_screen.js [hunk 7]
```diff
@@ -23,4 +23,32 @@ patch(PaymentScreen.prototype, {
...
+                (pi) => pi.payment_method_id.use_payment_terminal === "razorpay"
+            );
+            const transactions_ids = this.currentOrder.payment_ids.map((pi) => pi.transaction_id);
+            const res = super.addNewPaymentLine(paymentMethod);
+            const newPaymentLine = this.paymentLines.at(-1);
+            if (
+                res &&
+                (razorpayPaymentlines.length !== 1 ||
+                    Math.abs(newPaymentLine.amount) < razorpayPaymentlines[0].amount ||
+                    newPaymentLine.amount === 0 ||
```
Comment (adgu-odoo) [inline addons/pos_razorpay/static/src/app/screens/payment_screen/payment_screen.js:39]: In which case does this line triggers a true and the line above triggers a false ?

File: addons/pos_razorpay/static/src/app/screens/payment_screen/payment_screen.js [hunk 8]
```diff
@@ -23,4 +23,32 @@ patch(PaymentScreen.prototype, {
...
+                    transactions_ids.includes(newPaymentLine.transaction_id))
+            ) {
+                this.deletePaymentLine(newPaymentLine.uuid);
```
Comment (adgu-odoo) [inline addons/pos_razorpay/static/src/app/screens/payment_screen/payment_screen.js:42]: If you delete a line, I think you should tell the user why

File: addons/pos_razorpay/static/src/app/services/pos_store.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+patch(PosStore.prototype, {
+    async pay() {
+        const currentOrder = this.getOrder();
+        const refundedOrder = currentOrder?.lines[0]?.refunded_orderline_id?.order_id;
+        const razorpayPaymentlines = refundedOrder?.payment_ids.filter(
+            (pi) => pi.payment_method_id.use_payment_terminal === "razorpay"
+        );
+        if (refundedOrder && razorpayPaymentlines?.length === 1) {
+            await super.pay();
```
Comment (adgu-odoo) [inline addons/pos_razorpay/static/src/app/services/pos_store.js:12]:
```suggestion
        await super.pay();
        if (refundedOrder && razorpayPaymentlines?.length === 1) {
```
If you call super outside the if, you can remove the else under

File: addons/pos_razorpay/static/src/app/services/pos_store.js [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+            const paymentIds = refundedOrder.payment_ids || [];
+            // Add all the available payment lines in the refunded order if the current order amount is the same as the refunded order
+            if (Math.abs(currentOrder.getTotalDue()) === refundedOrder.amount_total) {
+                paymentIds.forEach((pi) => {
+                    if (pi.payment_method_id) {
```
Comment (adgu-odoo) [inline addons/pos_razorpay/static/src/app/services/pos_store.js:17]: what if the payment method is not available anymore in the config

File: addons/pos_razorpay/static/src/app/services/pos_store.js [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+                        const paymentLine = currentOrder.addPaymentline(pi.payment_method_id);
+                        paymentLine.setAmount(-pi.amount);
+                        paymentLine.updateRefundPaymentLine(pi);
+                    }
+                });
+            } else {
+                // Add available payment lines of refunded order based on conditions.
+                var getTotalDue = currentOrder.getTotalDue();
```
Comment (adgu-odoo) [inline addons/pos_razorpay/static/src/app/services/pos_store.js:25]:
```suggestion
                const getTotalDue = currentOrder.getTotalDue();
```

File: addons/pos_razorpay/static/src/app/services/pos_store.js [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+                if (getTotalDue < 0 && Math.abs(getTotalDue) > razorpayPaymentlines[0].amount) {
+                    const paymentLine = currentOrder.addPaymentline(
```
Comment (adgu-odoo) [inline addons/pos_razorpay/static/src/app/services/pos_store.js:27]: what if razorpay is not available anymore in the config

File: addons/pos_razorpay/static/src/app/services/pos_store.js [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+                        razorpayPaymentlines[0].payment_method_id
+                    );
+                    paymentLine.setAmount(-razorpayPaymentlines[0].amount);
+                    paymentLine.updateRefundPaymentLine(razorpayPaymentlines[0]);
+                }
+                if (currentOrder.getDue() < 0) {
+                    paymentIds.forEach((pi) => {
+                        var current_due = currentOrder.getDue();
```
Comment (adgu-odoo) [inline addons/pos_razorpay/static/src/app/services/pos_store.js:35]:
```suggestion
                        const current_due = currentOrder.getDue();
```

File: addons/pos_razorpay/static/src/app/services/pos_store.js [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+                        if (
+                            current_due < 0 &&
```
Comment (adgu-odoo) [inline addons/pos_razorpay/static/src/app/services/pos_store.js:37]: why do you check this, after the first payment_method that passes the two other conditions, the current_due will be 0 won't it ?

Comment (adgu-odoo) [reply addons/pos_razorpay/static/src/app/services/pos_store.js:37]: Also why do you add the current_due to a payment_method and not the amount that was paid with this payment method ?

File: addons/pos_razorpay/static/src/app/utls/payment/payment_razorpay.js [hunk 15]
```diff
@@ -86,11 +86,51 @@ export class PaymentRazorpay extends PaymentInterface {
...
     }
 
-    _processRazorpay(cid) {
+    _razorpayHandleRefundResponse(response) {
+        const paymentLine = this.pendingRazorpayline();
+        if (response?.error) {
+            paymentLine.setPaymentStatus("retry");
+            this._showError(response.error);
+            this._removePaymentHandler(["referenceId"]);
+            return Promise.resolve(false);
```
Comment (adgu-odoo) [inline addons/pos_razorpay/static/src/app/utls/payment/payment_razorpay.js:95]: do you need a Promise.resolve(false) here or a `return false;` is good ?

**Review by adgu-odoo — CHANGES_REQUESTED** [review]

File: addons/pos_razorpay/models/razorpay_pos_request.py [hunk 16]
```diff
@@ -29,10 +33,9 @@ def _call_razorpay(self, endpoint, payload):
...
-        request_timeout = self.payment_method.env['ir.config_parameter'].sudo().get_param('pos_razorpay.timeout', REQUEST_TIMEOUT)
+        endpoint = f'{self._razorpay_get_endpoint(endpoint)}{endpoint}'
```
Comment (Ysoroko) [inline addons/pos_razorpay/models/razorpay_pos_request.py:36]: the variable names need to be adapted, we are using `endpoint` after `endpoint` and assigning it to `endpoint`

**Review by Ysoroko — CHANGES_REQUESTED** [review]

File: addons/pos_razorpay/static/src/app/screens/payment_screen/payment_screen.js (hunk 7 above)
Comment (paan-odoo) [reply addons/pos_razorpay/static/src/app/screens/payment_screen/payment_screen.js:39]: If a payment line for a refund is added, the response (res) will be true. However, if the refunded order contains a single Razorpay payment line, and the amount of the new payment line is greater than the amount of the Razorpay payment line in the refunded order, while the new payment line amount is not equal to 0, the second condition will be false. In this case, there is no need to delete the added payment line.

Basically, we do not allow the addition of a new payment line if the above conditions are true.

File: addons/pos_razorpay/static/src/app/services/pos_store.js (hunk 14 above)
Comment (paan-odoo) [reply addons/pos_razorpay/static/src/app/services/pos_store.js:37]: > why do you check this, after the first payment_method that passes the two other conditions, the current_due will be 0 won't it ?

The current_due will change after adding a payment line, so it needs to be checked each time a payment line is added.

File: addons/pos_razorpay/models/pos_payment_method.py [hunk 17]
```diff
@@ -16,6 +16,40 @@ class PosPaymentMethod(models.Model):
...
+                'txnId': response.get('txnId'),
+                'paymentMode': response.get('paymentMode'),
+                'paymentCardType': response.get('paymentCardType'),
+                'paymentCardBrand': response.get('paymentCardBrand'),
+                'nameOnCard': response.get('nameOnCard'),
+                'acquirerCode': response.get('acquirerCode'),
+                'postingDate': response.get('postingDate'),
+            }
+        default_error_msg = _('The Razorpay POS refund request has encountered an unexpected error code.')
+        error = response.get('errorMessage') or default_error_msg
```
[the remark this answers is not in the document]
Comment (paan-odoo) [reply addons/pos_razorpay/models/pos_payment_method.py:50]: We are uncertain about the response because the card-related values may not be present for the `UPI` payment mode. Therefore, we are using `.get` to access the data safely.

**Comment by paan-odoo** [comment]
Hello @adgu-odoo @Ysoroko 

Suggested changes are done.

Thanks!

File: addons/point_of_sale/static/src/app/screens/payment_screen/payment_lines/payment_lines.xml [hunk 18]
```diff
@@ -56,7 +56,7 @@
                                     <div class="electronic_status">
                                         Waiting for card
                                     </div>
-                                    <div class="button send_payment_cancel" title="Cancel Payment Request" t-on-click="() => this.props.sendPaymentCancel(line)">
+                                    <div t-if="!this.props.isRefundOrder" class="button send_payment_cancel" title="Cancel Payment Request" t-on-click="() => this.props.sendPaymentCancel(line)">
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/screens/payment_screen/payment_lines/payment_lines.xml:59]: can't we cancel refund payment request?

Comment (paan-odoo) [reply addons/point_of_sale/static/src/app/screens/payment_screen/payment_lines/payment_lines.xml:59]: No, there is no request to cancel a refund/void request because refund/void approval comes instantly no need to wait.

File: addons/pos_razorpay/static/src/app/screens/payment_screen/payment_screen.js [hunk 19]
```diff
@@ -23,4 +24,36 @@ patch(PaymentScreen.prototype, {
             }
         });
     },
+    addNewPaymentLine(paymentMethod) {
+        if (paymentMethod.use_payment_terminal === "razorpay" && this.isRefundOrder) {
+            const refundedOrder = this.currentOrder.lines[0]?.refunded_orderline_id?.order_id;
+            const razorpayPaymentlines = refundedOrder.payment_ids.filter(
+                (pi) => pi.payment_method_id.use_payment_terminal === "razorpay"
+            );
+            const res = super.addNewPaymentLine(paymentMethod);
```
Comment (caburj) [inline addons/pos_razorpay/static/src/app/screens/payment_screen/payment_screen.js:33]:
```suggestion
            const res = await super.addNewPaymentLine(paymentMethod);
```

the super call returns a promise.

File: addons/point_of_sale/static/src/app/models/pos_payment.js [hunk 20]
```diff
@@ -86,6 +86,10 @@ export class PosPayment extends Base {
         }
         return isPaymentSuccessful;
     }
+
+    updateRefundPaymentLine(refundedPaymentLine) {
+        this.transaction_id = refundedPaymentLine.transaction_id;
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/models/pos_payment.js:91]: Why is the transaction_id of the refund payment the same as the refunded payment?

File: addons/pos_razorpay/static/src/app/screens/payment_screen/payment_screen.js [hunk 21]
```diff
@@ -23,4 +24,36 @@ patch(PaymentScreen.prototype, {
...
+                        "Adding a new Razorpay payment line is not allowed under the current conditions."
+                    ),
+                    { type: "warning", sticky: false }
+                );
+                return false;
+            }
+            if (res) {
+                newPaymentLine.setAmount(-razorpayPaymentlines[0].amount);
+                newPaymentLine.updateRefundPaymentLine(razorpayPaymentlines[0]);
+            }
```
Comment (caburj) [inline addons/pos_razorpay/static/src/app/screens/payment_screen/payment_screen.js:53]: I don't understand what these lines try to accomplish. The newly added razor payment line already has the correct amount. Perhaps it should only do `updateRefundPaymentLine`.

If your intention is to prevent addition of new line if there is already a pending refund, then that should be the code:

```js
if (this.hasPendingRefundPaymentLine()) {
    // don't add a line by NOT calling super
    // and perhaps show a warning notification.
}
```

File: addons/point_of_sale/static/src/app/models/pos_payment.js (hunk 20 above)
Comment (paan-odoo) [reply addons/point_of_sale/static/src/app/models/pos_payment.js:91]: The `transaction_id` is required for refund requests. Therefore, we first store `refundedPaymentLine.transaction_id` in the payment line. Once the refund response is received, we update it with the new `transaction_id`.

File: addons/pos_razorpay/static/src/app/utls/payment/payment_razorpay.js [hunk 22]
```diff
@@ -210,6 +288,22 @@ export class PaymentRazorpay extends PaymentInterface {
...
         });
     }
+
+    async _confirmVoidPayment() {
+        return new Promise((resolve, reject) => {
+            this.env.services.dialog.add(ConfirmationDialog, {
+                title: _t("Void Payment Confirmation"),
+                body: _t(
+                    "Your transaction isn't settled yet, and the refund is less than the amount paid.\n" +
+                        "Do you want to cancel the entire transaction?"
```
Comment (caburj) [inline addons/pos_razorpay/static/src/app/utls/payment/payment_razorpay.js:298]:
```suggestion
                        "The full amount will be cancelled, do you want to proceed?"
```

**Review by caburj — APPROVED** [review]
robodoo r+
